### ISSUE TITLE
Using Spring Cloud Sleuth with OpenZipkin. GUI not showing all spans

### ISSUE THREAD

**Issue by Mrc0113** [spring-cloud/spring-cloud-sleuth#1916]
I am trying to prototype getting Spring Cloud Sleuth working with Spring Cloud Stream microservices and sending tracing info to OpenZipkin. It seems to be working however for some reason I'm not seeing what I'd expect in the OpenZipkin GUI. 

I uploaded my code here and it should only take 5-10 minutes to get completely running. Including the broker + openzipkin. https://github.com/Mrc0113/spring-cloud-stream-sleuth

Versions: 
* I've tried the latest release versions (2020.0.2) and the latest snapshot versions (2020.0.3-SNAPSHOT) of spring cloud

Expected Flow: Try-Me Publisher -> uppercase-in-0 topic on broker -> Uppercase Microservice (Trace starts here) -> uppercase-out-0 topic -> Reverse Microservice (Tracing still enabled here...same traceid, different span id) -> reverse-out-0 topic -> Try Me Consumer if you want. 

In the OpenZipkin GUI I expected to see the trace showing Spans for both the "UppercaseApp" and the "ReverseApp" however it just shows the spans related to the "UppercaseApp". Shouldn't it show the "ReverseApp" as well?  

When you download the JSON from OpenZipkin it shows 6 spans including both microservices so I'm not sure if I'm doing something wrong or if this is just a bug in OpenZipkin's GUI. 

[![openzipkinimage][1]][1]

  [1]: https://i.stack.imgur.com/kHNLt.png

Downloaded JSON:
```
[
  {
    "traceId": "ff03e5a7078c7300",
    "parentId": "ff03e5a7078c7300",
    "id": "0e954000ac123168",
    "kind": "CONSUMER",
    "timestamp": 1617974110519267,
    "duration": 17,
    "localEndpoint": {
      "serviceName": "uppercaseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase"
    }
  },
  {
    "traceId": "ff03e5a7078c7300",
    "parentId": "0e954000ac123168",
    "id": "7cc8f2fe19e3e429",
    "name": "handle",
    "timestamp": 1617974110527406,
    "duration": 3800,
    "localEndpoint": {
      "serviceName": "uppercaseapp",
      "ipv4": "192.168.1.25"
    }
  },
  {
    "traceId": "ff03e5a7078c7300",
    "parentId": "0e954000ac123168",
    "id": "b136a2d5e017b1c6",
    "kind": "PRODUCER",
    "name": "send",
    "timestamp": 1617974110532062,
    "duration": 1802,
    "localEndpoint": {
      "serviceName": "uppercaseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase"
    }
  },
  {
    "traceId": "ff03e5a7078c7300",
    "parentId": "8432f0d20ee1c58a",
    "id": "eb4d8e0d2265ccf9",
    "kind": "CONSUMER",
    "timestamp": 1617974110633526,
    "duration": 23,
    "localEndpoint": {
      "serviceName": "reverseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase-out-0"
    }
  },
  {
    "traceId": "ff03e5a7078c7300",
    "parentId": "eb4d8e0d2265ccf9",
    "id": "5d342c6ff6cda3f9",
    "name": "handle",
    "timestamp": 1617974110643065,
    "duration": 4683,
    "localEndpoint": {
      "serviceName": "reverseapp",
      "ipv4": "192.168.1.25"
    }
  },
  {
    "traceId": "ff03e5a7078c7300",
    "parentId": "eb4d8e0d2265ccf9",
    "id": "9d467b04bf9a832e",
    "kind": "PRODUCER",
    "name": "send",
    "timestamp": 1617974110649193,
    "duration": 1998,
    "localEndpoint": {
      "serviceName": "reverseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase-out-0"
    }
  }
]
```

**Comment by jonatan-ivanov** [comment]
@Mrc0113 Taking a look at the output, the first span of the `reverseapp` might cause the issue:
```
    "traceId": "ff03e5a7078c7300",
    "parentId": "8432f0d20ee1c58a",
    "id": "eb4d8e0d2265ccf9",
```

It's `parentId` (`8432f0d20ee1c58a`) does not seem to point to any other spans, could you pease double check a few things?
1. Could you please check the logs if you see any spans with an id of the `parentId` of the first span of `reverseapp`?
2. Could you please check if the message headers contain the right tracing information?

There are scenarios around spring-cloud-stream that are not supported, this might be one of them.
@olegz Do you happen to know if this should work?

**Comment by Mrc0113** [comment]
Hi @jonatan-ivanov, 

Thanks for looking into this! @olegz I think @jonatan-ivanov is correct in that it seems that Spring Cloud Stream doesn't use the incoming spanid as any sort of parent. It seems to only use the traceid. Maybe we should have a quick session with @marcingrzejszczak to discuss what should happen at some point? 

Here is what I'm seeing: 
Remember: Publisher -> Broker -> Uppercase SCSt. Microservice -> Broker -> Reverse SCSt. Microservice -> Broker 

## Updated info sent to Zipkin
```
[
  {
    "traceId": "9ff1228f1a8a4e67",
    "parentId": "9ff1228f1a8a4e67",
    "id": "dbf76d900a14112c",
    "kind": "CONSUMER",
    "timestamp": 1619545083351152,
    "duration": 19,
    "localEndpoint": {
      "serviceName": "uppercaseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase"
    }
  },
  {
    "traceId": "9ff1228f1a8a4e67",
    "parentId": "dbf76d900a14112c",
    "id": "6e3ecf9fc5c6d707",
    "name": "handle",
    "timestamp": 1619545083358879,
    "duration": 6290,
    "localEndpoint": {
      "serviceName": "uppercaseapp",
      "ipv4": "192.168.1.25"
    }
  },
  {
    "traceId": "9ff1228f1a8a4e67",
    "parentId": "dbf76d900a14112c",
    "id": "ad1695a6eaba2137",
    "kind": "PRODUCER",
    "name": "send",
    "timestamp": 1619545083366144,
    "duration": 3001,
    "localEndpoint": {
      "serviceName": "uppercaseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase"
    }
  },
  {
    "traceId": "9ff1228f1a8a4e67",
    "parentId": "e9fe9a6605e23878",
    "id": "c6f2d973203509d1",
    "kind": "CONSUMER",
    "timestamp": 1619545083468677,
    "duration": 20,
    "localEndpoint": {
      "serviceName": "reverseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase-out-0"
    }
  },
  {
    "traceId": "9ff1228f1a8a4e67",
    "parentId": "c6f2d973203509d1",
    "id": "cda72cef9480aa21",
    "name": "handle",
    "timestamp": 1619545083477555,
    "duration": 5914,
    "localEndpoint": {
      "serviceName": "reverseapp",
      "ipv4": "192.168.1.25"
    }
  },
  {
    "traceId": "9ff1228f1a8a4e67",
    "parentId": "c6f2d973203509d1",
    "id": "54af921fae57372c",
    "kind": "PRODUCER",
    "name": "send",
    "timestamp": 1619545083484385,
    "duration": 1788,
    "localEndpoint": {
      "serviceName": "reverseapp",
      "ipv4": "192.168.1.25"
    },
    "remoteEndpoint": {
      "serviceName": "broker"
    },
    "tags": {
      "channel": "uppercase-out-0"
    }
  }
]
```

## Message being output from uppercase microservice (looking at it on the broker) has these headers which I believe are related:
*   Key 'b3' (String): 9ff1228f1a8a4e67-ad1695a6eaba2137-1
*   Key 'nativeHeaders' (String): rO0ABXNyACxvcmcuc3ByaW5nZnJhbWV3b3JrLnV0aWwuTGlua2VkTXVsdGlWYWx1ZU1hcDTATlwQbMD7AgAAeHIALW9yZy5zcHJpbmdmcmFtZXdvcmsudXRpbC5NdWx0aVZhbHVlTWFwQWRhcHRlcu38oXoIPvU2AgABTAAJdGFyZ2V0TWFwdAAPTGphdmEvdXRpbC9NYXA7eHBzcgAXamF2YS51dGlsLkxpbmtlZEhhc2hNYXA0wE5cEGzA+wIAAVoAC2FjY2Vzc09yZGVyeHIAEWphdmEudXRpbC5IYXNoTWFwBQfawcMWYNEDAAJGAApsb2FkRmFjdG9ySQAJdGhyZXNob2xkeHA/QAAAAAAADHcIAAAAEAAAAAF0AAJiM3NyACNqYXZhLnV0aWwuQ29sbGVjdGlvbnMkU2luZ2xldG9uTGlzdCrvKRA8p5uXAgABTAAHZWxlbWVudHQAEkxqYXZhL2xhbmcvT2JqZWN0O3hwdAAjOWZmMTIyOGYxYThhNGU2Ny1hZDE2OTVhNmVhYmEyMTM3LTF4AA==

## Spring Message in the reverse microservice by the team it gets handed to function with these headers:
*  org.springframework.cloud.sleuth.Span=RealSpan(9ff1228f1a8a4e67/cda72cef9480aa21) 
*  traceHandlerParentSpan=RealSpan(9ff1228f1a8a4e67/c6f2d973203509d1)
*  nativeHeaders={}

If I'm following what @jonatan-ivanov is saying (and I'm no tracing expert) then we should probably see `ad1695a6eaba2137` as the parent id? not sure where `c6f2d973203509d1` is coming from. 


Tangential side note for @olegz:
* It would be nice to be able to provide channel information as part of the trace in some way instead of just "broker". If you have a bunch of event-driven microservices participating in choreography it would just show each app communicating with the broker which is not super useful on it's own.

**Comment by jonatan-ivanov** [comment]
@Mrc0113 This is my understanding. The parentId of the span in the consumer should be the id of the span in the producer.
I need to debug this (thank you for the sample project): we need to check what happens between the message is received and your `Function` is called. My guess is the context is not propagated (as I mentioned above, this might be an unsupported scenario).

**Comment by Mrc0113** [comment]
Hi @jonatan-ivanov, glad the project will come in useful and thanks for looking into this. Sorry for the delayed response, I completely missed the notification that you had responded 🤦‍♂️

**Comment by marcingrzejszczak** [comment]
Hi I've used the Spring Cloud version 2020.0.4 and Boot 2.4.10 and checked out your sample. I got the following result

![image](https://user-images.githubusercontent.com/3297437/135826809-af2eaa93-84ce-4010-98fc-f71972c8b63f.png)

logs

.Uppercase-app
```
2021-10-04 11:23:09.996 DEBUG [UppercaseApp,,] 25059 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Invoking function uppercase<class java.lang.String, class java.lang.String>
2021-10-04 11:23:09.997 DEBUG [UppercaseApp,,] 25059 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Executing around advise(s): org.springframework.cloud.sleuth.instrument.messaging.TraceFunctionAroundWrapper@61ff6a49
2021-10-04 11:23:10.017 DEBUG [UppercaseApp,960b59a1f712189b,746f158ddcdd03ad] 25059 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Invoking function uppercase<class java.lang.String, class java.lang.String>
2021-10-04 11:23:10.020 DEBUG [UppercaseApp,960b59a1f712189b,746f158ddcdd03ad] 25059 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Converted Message: GenericMessage [payload=byte[12], headers={solace_senderTimestamp=1633339389951, solace_expiration=0, deliveryAttempt=1, solace_destination=uppercase-in-0, solace_timeToLive=0, solace_receiveTimestamp=0, target-protocol=kafka, acknowledgmentCallback=com.solace.spring.cloud.stream.binder.util.JCSMPAcknowledgementCallbackFactory$JCSMPAcknowledgementCallback@483cb4e9, solace_discardIndication=false, org.springframework.cloud.sleuth.Span=RealSpan(960b59a1f712189b/746f158ddcdd03ad), solace_dmqEligible=false, solace_priority=-1, traceHandlerParentSpan=RealSpan(960b59a1f712189b/960b59a1f712189b), solace_redelivered=false, id=4daaa397-5433-c22a-a949-a9e16adf4f09, contentType=application/json, solace_senderId=Try-Me-Pub/solclientjs/firefox-92.0.0-Linux-0.0.0/1361791701/0001}] to: org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$OriginalMessageHolder@c3fe06f
2021-10-04 11:23:10.020  INFO [UppercaseApp,960b59a1f712189b,746f158ddcdd03ad] 25059 --- [pool-5-thread-1] s.s.s.SpringCloudStreamSleuthApplication : Received: Hello world!
```

.Reverse app

```

2021-10-04 11:23:10.112 DEBUG [ReverseApp,,] 25348 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Invoking function reverse<class java.lang.String, class java.lang.String>
2021-10-04 11:23:10.113 DEBUG [ReverseApp,,] 25348 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Executing around advise(s): org.springframework.cloud.sleuth.instrument.messaging.TraceFunctionAroundWrapper@7a682d35
2021-10-04 11:23:10.139 DEBUG [ReverseApp,960b59a1f712189b,f224f79ec7a3af15] 25348 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Invoking function reverse<class java.lang.String, class java.lang.String>
2021-10-04 11:23:10.140 DEBUG [ReverseApp,960b59a1f712189b,f224f79ec7a3af15] 25348 --- [pool-5-thread-1] c.f.c.c.BeanFactoryAwareFunctionRegistry : Converted Message: GenericMessage [payload=byte[12], headers={solace_senderTimestamp=1633339389951, solace_scst_messageVersion=1, solace_expiration=0, deliveryAttempt=2, solace_destination=uppercase-out-0, solace_timeToLive=0, solace_receiveTimestamp=0, target-protocol=kafka, nativeHeaders={}, acknowledgmentCallback=com.solace.spring.cloud.stream.binder.util.JCSMPAcknowledgementCallbackFactory$JCSMPAcknowledgementCallback@1d89b71, solace_discardIndication=false, org.springframework.cloud.sleuth.Span=RealSpan(960b59a1f712189b/f224f79ec7a3af15), solace_dmqEligible=false, source-type=kafka, solace_priority=4, traceHandlerParentSpan=RealSpan(960b59a1f712189b/d9567520bb69e187), solace_redelivered=false, id=260d3f6f-b992-a423-3844-8db4bd470ba0, contentType=application/json, solace_senderId=Try-Me-Pub/solclientjs/firefox-92.0.0-Linux-0.0.0/1361791701/0001}] to: org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry$OriginalMessageHolder@259ff46d
2021-10-04 11:23:10.140  INFO [ReverseApp,960b59a1f712189b,f224f79ec7a3af15] 25348 --- [pool-5-thread-1] s.s.s.SpringCloudStreamSleuthApplication : Received: HELLO WORLD!
```

and zipkin

![image](https://user-images.githubusercontent.com/3297437/135827089-b2142886-a053-4fec-8ac5-71cdb2ada700.png)

It looks good IMO, or am I missing sth?

**Comment by Mrc0113** [comment]
awesome @marcingrzejszczak! I never tried it since there was no comment on this issue, but maybe the issue was resolved as that does look better! I should be able to give it a try tomorrow and I'll update this ticket.

**Comment by Mrc0113** [comment]
okay I can verify that this issue is now resolved when using Spring Boot v2.4.10 and Spring Cloud v2020.0.4. 
Thanks!
